scipy/stats/_distribution_infrastructure.py

```diff
@@ -0,0 +1,5217 @@
+
+    ## Algorithms
+
+    def _quadrature(self, integrand, limits=None, args=None,
+                    params=None, log=False):
+        # Performs numerical integration of an integrand between limits.
+        # Much of this should be added to `_tanhsinh`.
+        a, b = self._support(**params) if limits is None else limits
+        a, b = np.broadcast_arrays(a, b)
+        if not a.size:
+            # maybe need to figure out result type from a, b
+            return np.empty(a.shape, dtype=self._dtype)
+        args = [] if args is None else args
+        params = {} if params is None else params
+        f, args = _kwargs2args(integrand, args=args, kwargs=params)
+        args = np.broadcast_arrays(*args)
+        # If we know the median or mean, consider breaking up the interval
+        rtol = None if _isnull(self.tol) else self.tol
+        res = _tanhsinh(f, a, b, args=args, log=log, rtol=rtol)
+        # For now, we ignore the status, but I want to return the error
+        # estimate - see question 5 at the top.
+        return res.integral
+
+    def _solve_bounded(self, f, p, *, bounds=None, params=None):
+        # Finds the argument of a function that produces the desired output.
+        # Much of this should be added to _bracket_root / _chandrupatla.
+        xmin, xmax = self._support(**params) if bounds is None else bounds
+        params = {} if params is None else params
+
+        p, xmin, xmax = np.broadcast_arrays(p, xmin, xmax)
+        if not p.size:
+            # might need to figure out result type based on p
+            return np.empty(p.shape, dtype=self._dtype)
+
+        def f2(x, p, **kwargs):
+            return f(x, **kwargs) - p
+
+        f3, args = _kwargs2args(f2, args=[p], kwargs=params)
+        # If we know the median or mean, should use it
+
+        # Any operations between 0d array and a scalar produces a scalar, so...
+        shape = xmin.shape
+        xmin, xmax = np.atleast_1d(xmin, xmax)
+
+        a = -np.ones_like(xmin)
+        b = np.ones_like(xmax)
+
+        i = np.isfinite(xmin) & np.isfinite(xmax)
+        a[i] = xmin[i]
+        b[i] = xmax[i]
+
+        i = np.isfinite(xmin) & ~np.isfinite(xmax)
+        a[i] = xmin[i]
+        b[i] = xmin[i] + 1
+
+        i = np.isfinite(xmax) & ~np.isfinite(xmin)
+        a[i] = xmax[i] - 1
+        b[i] = xmax[i]
+
+        xmin = xmin.reshape(shape)
+        xmax = xmax.reshape(shape)
+        a = a.reshape(shape)
+        b = b.reshape(shape)
+
+        res = _bracket_root(f3, xl0=a, xr0=b, xmin=xmin, xmax=xmax, args=args)
+        # For now, we ignore the status, but I want to use the bracket width
+        # as an error estimate - see question 5 at the top.
+        xrtol = None if _isnull(self.tol) else self.tol
+        return _chandrupatla(f3, a=res.xl, b=res.xr, args=args, xrtol=xrtol).x
+
+    ## Other
+
+    def _overrides(self, method_name):
+        # Determines whether a class overrides a specified method.
```

Comment:
> but this is typically one of several considerations.

I'm still slightly confused to be honest. The change in logic from the old distribution framework doesn't seem large, since the docs below say:

_Typically, a formula is always used if available. We fall back to "log/exp" if a formula for the logarithm or exponential of the quantity is available, and we use quadrature otherwise._

That I understand pretty easily, since an explicit formula is always going to be more accurate and faster, and numerical integration is always going to be the slow fallback of last resort. What I still have trouble understanding is how/where the "depends on error tolerance given by the user" comes in (I think that deviates from the sentence in italics above?) and hence why the new structure is necessary.

Reply:
Sorry, I realized that my initial response was not on the mark. Consider 

https://github.com/scipy/scipy/blob/4979e7579f37060eaeaf05a7d622a4e7dd6b13a7/scipy/stats/_distribution_infrastructure.py#L3207-L3216

The first `_override` is to check whether there is an explicit formula for the log CDF. This is the one that could conceivably be eliminated if we simply had a public `logcdf` and a single, private `_logcdf` that the developer can override. But before getting into that, let's look at the other two uses of `_override` in that function:

If the user hasn't specified a tolerance requirement and a formula for CDF is available (`self._overrides('_cdf_formula')`), it's OK to just take the `log` of the CDF. If the user has specified a tolerance requirement and a formula for the log CCDF is available (`self._overrides('_logccdf_formula')`), then the extra time required to take a "logarithmic complement" is warranted to better satisfy the tolerance. If neither formula is available, then we would have to use quadrature to compute those quantities, so it's best to just use quadrature to compute the quantity we want directly. This may not be the best possible logic, but the specific logic a detail that can (and will) be improved.

As for why we don't simply allow the developer to replace a single `_logcdf`, the biggest reason is that although we currently use the formula if it is provided, the formula is *not* always going to be the most accurate result available. For instance, although it's not implemented set up to do this yet, something we might want to do is allow the `_logcdf_formula` method to fall back to `_logcdf_dispatch` if the argument is in a region it's not designed for. We do this sort of thing in the `_moment_xyz_dispatch` methods (although there, there is a default `_formula` method that just returns `None`).

> Some explanation of why it's done this way would be helpful I think.

I will add a comment in the code to document this.

Reply:
Thanks for the detailed explanation. The `_isnull(self.tol)` part feels intuitively wrong to me, since it's switching calculation methods (independent if the user said `tol=0.1` or `tol=1e-10`) which is the kind of thing that usually leads to surprises.

However, that's just based on general principles. I don't have time to dig in further, so I'm happy to go with your judgement - if you add a comment that addresses my initial confusion then please feel free to resolve this comment.

Reply:
I agree; that is currently a placeholder for better logic. I'll add the comment. Thanks!

Reply:
Done in next commit.